modules/xml/xml-grammar.ym

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2017 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+%code top {
+#include "xml-parser.h"
+}
+
+
+%code {
+#include "xml.h"
+#include "cfg-parser.h"
+#include "cfg-grammar.h"
+#include "xml-grammar.h"
+#include "syslog-names.h"
+#include "messages.h"
+#include "plugin.h"
+
+#include <string.h>
+
+}
+
+%name-prefix "xml_"
+
+/* this parameter is needed in order to instruct bison to use a complete
+ * argument list for yylex/yyerror */
+
+%lex-param {CfgLexer *lexer}
+%parse-param {CfgLexer *lexer}
+%parse-param {LogParser **instance}
+%parse-param {gpointer arg}
+
+/* INCLUDE_DECLS */
+
+%token KW_XML
+%token KW_PREFIX
+%token KW_DROP_INVALID
+%token KW_EXCLUDE_TAGS
+%token KW_STRIP_WHITESPACES
+
+%type	<ptr> parser_xml
+%%
+
+start
+  : LL_CONTEXT_PARSER parser_xml          { YYACCEPT; }
+  ;
+
+parser_xml
+  : KW_XML '('
+     {
+        last_parser = *instance = (LogParser *) xml_parser_new(configuration);
+     }
+
+        xml_options
+                       ')'
+     {
+        $$ = last_parser;
+     }
+  ;
+
+xml_options
+        : xml_opt xml_options
+        |
+        ;
+
+xml_opt
+        : KW_PREFIX '(' string ')'      { xml_parser_set_prefix(last_parser, $3); free($3); }
+        | KW_DROP_INVALID '(' yesno ')' { xml_parser_set_forward_invalid(last_parser, !$3); };
```

Comment:
There is a similar problem/behaviour/option in dbparser, where we added KW_DROP_UNMATCHED. I wanted to align my option with that. That's why I used drop invalid here. Probably subjective feeling, but I think in case of an invalid input, the term 'drop' comes earlier to one's mind than 'forward'.
From coding point of view however, our internal api (parser_process) thinks in "forwarding", so it was easier to work with the positive version it internally. So I use this negation between the user api and the code internals.

Reply:
It's true, but if the default value is yes, then a user want to set this option only if she or he wants to forward.

Reply:
We decided to have drop_invalid: no as default value.